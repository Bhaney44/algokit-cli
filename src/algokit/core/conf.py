import logging
import os
import platform
import sys
from importlib import metadata
from pathlib import Path

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib

from algokit.core.constants import ALGOKIT_CONFIG

PACKAGE_NAME = "algokit"

logger = logging.getLogger(__name__)


def get_app_config_dir() -> Path:
    """Get the application config files location - things that should persist, and potentially follow a user"""
    os_type = platform.system().lower()
    if os_type == "windows":
        config_dir = os.getenv("APPDATA")
    else:
        config_dir = os.getenv("XDG_CONFIG_HOME")
    if config_dir is None:
        config_dir = "~/.config"
    return _get_relative_app_path(config_dir)


def get_app_state_dir() -> Path:
    """Get the application state files location - things the user wouldn't normally interact with directly"""
    os_type = platform.system().lower()
    if os_type == "windows":
        state_dir = os.getenv("LOCALAPPDATA")
    elif os_type == "darwin":
        state_dir = "~/Library/Application Support"
    else:
        state_dir = os.getenv("XDG_STATE_HOME")
    if state_dir is None:
        state_dir = "~/.local/state"
    return _get_relative_app_path(state_dir)


def _get_relative_app_path(base_dir: str) -> Path:
    path = Path(base_dir).expanduser()
    result = path / PACKAGE_NAME
    result.mkdir(parents=True, exist_ok=True)
    # resolve path in case of UWP sandbox redirection
    return result.resolve()


def get_current_package_version() -> str:
    return metadata.version(PACKAGE_NAME)


def get_algokit_config(project_dir: Path, config_filename: str = ALGOKIT_CONFIG) -> dict | None:
    """
    Load and parse a TOML configuration file.
    :param project_dir: Project directory path.
    :param config_filename: Configuration filename.
    :return: A dictionary containing the configuration or None if not found.
    """
    try:
        config_path = project_dir / config_filename
        config_text = config_path.read_text("utf-8")
    except FileNotFoundError:
        logger.debug(f"No {config_filename} file found in the project directory.")
        return None
    except Exception as ex:
        logger.debug(f"Unexpected error reading {config_filename} file: {ex}")
        return None

    try:
        return tomllib.loads(config_text)
    except Exception as ex:
        logger.debug(f"Error parsing {config_filename} file: {ex}")
        return None
