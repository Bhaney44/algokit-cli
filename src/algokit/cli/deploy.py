import logging
import os
from pathlib import Path

import click

from algokit.core.deploy import DEPLOY_CONFIGS, execute_deploy_command, load_deploy_command, load_deploy_config

logger = logging.getLogger(__name__)


class HiddenSecret:
    def __init__(self, secret: str = ""):
        self.secret = secret

    def __str__(self) -> str:
        return "*" * len(self.secret)


DEPLOYER_KEY = "DEPLOYER_MNEMONIC"
DISPENSER_KEY = "DISPENSER_MNEMONIC"


@click.command("deploy")
@click.argument("network", type=click.Choice(list(DEPLOY_CONFIGS)), default="localnet", required=True)
@click.option(
    "--deployer_mnemonic",
    required=True,
    default=lambda: HiddenSecret(os.environ.get(DEPLOYER_KEY, "")),
    hide_input=True,
)
@click.option(
    "--dispenser_mnemonic",
    required=False,
    prompt=True,
    default=lambda: HiddenSecret(os.environ.get(DISPENSER_KEY, "")),
    hide_input=True,
)
@click.option(
    "--custom_deploy_command",
    type=str,
    default=None,
    help="Custom deploy command. If not provided, will load the deploy command from .algokit.toml file in the current directory.",  # noqa: E501
)
def deploy_command(network: str, deployer_mnemonic: str, dispenser_mnemonic: str, custom_deploy_command: str) -> None:
    """Deploy smart contracts from AlgoKit compliant repository.
    Will load the deploy command from .algokit.toml file in the current directory.
    Will load default configuration for network using AlgoNode. To override the default configuration,
    create a .env file in the current directory and set the required environment variables prefixed with the network name (e.g. LOCALNET_ALGOD_TOKEN and etc)
    """  # noqa: E501

    # Load deploy config
    project_dir = Path.cwd()
    deploy_config = load_deploy_config(network, project_dir)

    # Inject deploy config with mnemonic
    deploy_config[DEPLOYER_KEY] = deployer_mnemonic
    deploy_config[DISPENSER_KEY] = dispenser_mnemonic

    # Load deploy command
    deploy_command = custom_deploy_command or load_deploy_command(network_name=network, project_dir=Path.cwd())

    # Execute deploy command
    logger.info("Deploying smart contracts from AlgoKit compliant repository.")
    execute_deploy_command(command=deploy_command, deploy_config=deploy_config, project_dir=project_dir)
    logger.info("Deployment completed successfully ðŸŽ‰")
