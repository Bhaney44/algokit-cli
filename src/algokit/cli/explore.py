import logging
from urllib.parse import urlencode

import click

from algokit.core.constants import ALGORAND_NETWORKS, AlgorandNetworkConfiguration

logger = logging.getLogger(__name__)


def get_dappflow_url(network: AlgorandNetworkConfiguration) -> str:
    # Filter out empty values
    dappflow_config = {key.lower(): value for key, value in network.items() if value}

    # Map keys to dappflow keys
    dappflow_config["algod_url"] = dappflow_config.pop("algod_server")
    dappflow_config["indexer_url"] = dappflow_config.pop("indexer_server")

    # Encode query string
    query_string = urlencode(dappflow_config)

    # Return dappflow url
    return f"https://app.dappflow.org/setup-config?{query_string}"


@click.command("explore", help="Explore the specified network in the browser using Dappflow.")
@click.argument("network", type=click.Choice(list(ALGORAND_NETWORKS)), default="localnet", required=False)
def explore_command(network: str) -> None:
    url = get_dappflow_url(ALGORAND_NETWORKS[network])
    logger.info(f"Opening {network} in https://app.dappflow.org using default browser")
    click.launch(url)
