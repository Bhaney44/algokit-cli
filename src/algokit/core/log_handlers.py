import functools
import logging
import os
import sys
from logging.handlers import RotatingFileHandler
from types import TracebackType
from typing import Any, Callable, ParamSpec, TypeVar

import click
from click.globals import resolve_color_default

from .conf import get_app_state_dir

__all__ = [
    "initialise_logging",
    "output_options",
    "uncaught_exception_logging_handler",
    "EXTRA_EXCLUDE_FROM_CONSOLE",
    "EXTRA_EXCLUDE_FROM_LOGFILE",
    "get_colour_pref_from_args",
]


class ClickHandler(logging.Handler):
    """Handle console output with click.echo(...)

    Slightly special in that this class acts as both a sink and an additional formatter,
    but they're kind of intertwined for our use case of actually displaying things to the user.
    """

    styles: dict[str, dict[str, Any]] = {
        "critical": {"fg": "red", "bold": True},
        "error": {"fg": "red"},
        "warning": {"fg": "yellow"},
        "debug": {"fg": "cyan"},
    }

    def emit(self, record: logging.LogRecord) -> None:
        try:
            msg = self.format(record)
            level = record.levelname.lower()
            if level in self.styles:
                # if user hasn't disabled colors/styling, just use that
                if resolve_color_default() is not False:
                    level_style = self.styles[level]
                    msg = click.style(msg, **level_style)
                # otherwise, prefix the level name
                else:
                    msg = f"{level.upper()}: {msg}"
            click.echo(msg)
        except Exception:
            self.handleError(record)


class NoExceptionFormatter(logging.Formatter):
    """Prevent automatically displaying exception/traceback info.
    (without interfering with other formatters that might later want to add such information)
    """

    def formatException(self, *args: Any) -> str:  # noqa: N802
        return ""

    def formatStack(self, *args: Any) -> str:  # noqa: N802
        return ""


CONSOLE_LOG_HANDLER_NAME = "console_log_handler"

EXCLUDE_FROM_KEY = "exclude_from"
EXCLUDE_FROM_CONSOLE_VALUE = "console"
EXCLUDE_FROM_LOGFILE_VALUE = "logfile"

EXTRA_EXCLUDE_FROM_CONSOLE = {EXCLUDE_FROM_KEY: EXCLUDE_FROM_CONSOLE_VALUE}
EXTRA_EXCLUDE_FROM_LOGFILE = {EXCLUDE_FROM_KEY: EXCLUDE_FROM_LOGFILE_VALUE}


class ManualExclusionFilter(logging.Filter):
    def __init__(self, exclude_value: str):
        super().__init__()
        self.exclude_value = exclude_value

    def filter(self, record: logging.LogRecord) -> bool:
        return getattr(record, EXCLUDE_FROM_KEY, None) != self.exclude_value


# HACK: due to there being no support for "global flags" in click (see: https://github.com/pallets/click/issues/66),
#       we handle these flags specially. All groups/commands should have the @output_options decorator below added,
#       so that the help doc shows these as correctly being applicable anywhere, but in order to prevent user
#       confusion when specifying before or after a command gives different results, and similarly for colour options
#       (e.g. if you pass --no-color as the last flag on the command line, it's very surprising if colour shows up
#       in some of the output, which it can if there's output in parent groups)
_VERBOSE_FLAGS = ["--verbose", "-v"]


def initialise_logging() -> None:
    console_log_handler = ClickHandler()
    if set(_VERBOSE_FLAGS).intersection(sys.argv):
        console_log_level = logging.DEBUG
    else:
        console_log_level = logging.INFO
    console_log_handler.setLevel(console_log_level)
    console_log_handler.name = CONSOLE_LOG_HANDLER_NAME
    console_log_handler.formatter = NoExceptionFormatter()
    console_log_handler.addFilter(ManualExclusionFilter(exclude_value=EXCLUDE_FROM_CONSOLE_VALUE))

    file_log_handler = RotatingFileHandler(
        filename=get_app_state_dir() / "cli.log",
        maxBytes=1 * 1024 * 1024,
        backupCount=5,
    )
    file_log_handler.setLevel(logging.DEBUG)
    file_log_handler.formatter = logging.Formatter(
        "%(asctime)s.%(msecs)03d %(name)s %(levelname)s %(message)s", datefmt="%Y-%m-%dT%H:%M:%S"
    )
    file_log_handler.addFilter(ManualExclusionFilter(exclude_value=EXCLUDE_FROM_LOGFILE_VALUE))

    logging.basicConfig(level=logging.DEBUG, handlers=[console_log_handler, file_log_handler], force=True)


def uncaught_exception_logging_handler(
    exc_type: type[BaseException], exc_value: BaseException, exc_traceback: TracebackType | None
) -> None:
    """Function to be used as sys.excepthook, which logs uncaught exceptions."""
    if issubclass(exc_type, KeyboardInterrupt):
        # don't log ctrl-c or equivalents
        sys.__excepthook__(exc_type, exc_value, exc_traceback)
    else:
        logging.critical(f"Unhandled {exc_type.__name__}: {exc_value}", exc_info=(exc_type, exc_value, exc_traceback))


_ENABLE_COLOR_FLAG = "--color"
_DISABLE_COLOR_FLAG = "--no-color"


T = TypeVar("T")
P = ParamSpec("P")


def output_options(func: Callable[P, T]) -> Callable[P, T]:
    # if you're wondering why these options do nothing, please see the HACK comment above,
    # where the names are declared
    @click.option(
        *_VERBOSE_FLAGS,
        is_flag=True,
        expose_value=False,
        help="Enable logging of DEBUG messages to the console",
    )
    @click.option(
        f"{_ENABLE_COLOR_FLAG}/{_DISABLE_COLOR_FLAG}",
        default=None,
        expose_value=False,
        help="Force enable or disable of console output styling",
    )
    @functools.wraps(func)
    def wrapped(*args: P.args, **kwargs: P.kwargs) -> T:
        return func(*args, **kwargs)

    return wrapped


def get_colour_pref_from_args(argv: list[str]) -> bool | None:
    # search in reverse, last specified wins.
    # see HACK note above for why this function exists
    for arg in reversed(argv):
        if arg == _ENABLE_COLOR_FLAG:
            return True
        if arg == _DISABLE_COLOR_FLAG:
            return False
    # support NO_COLOR (ref: https://no-color.org) env var as default value,
    if os.getenv("NO_COLOR"):
        return False
    return None
