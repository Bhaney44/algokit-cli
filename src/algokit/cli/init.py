import logging
from pathlib import Path

import click
from algokit.core.log_handlers import output_options

logger = logging.getLogger(__name__)


@click.command("init", short_help="Initializes a new project.")
@click.argument("path", required=False, type=click.Path(resolve_path=True, path_type=Path))
@output_options
def init_command(path: Path | None) -> None:
    """Initializes a new project."""
    if path is None:
        logger.info("Initialising in current directory")
        path = Path.cwd()
    # TODO: the thing
    logger.info(f"Initialized the project in {click.format_filename(path)}")
