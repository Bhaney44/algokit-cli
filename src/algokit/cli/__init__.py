import click
from algokit.cli.init import init_command
from algokit.cli.sandbox import sandbox_group
from algokit.core.conf import PACKAGE_NAME
from algokit.core.log_handlers import output_options


@click.group(
    help="AlgoKit is your one-stop shop to develop applications on the Algorand blockchain.",
    context_settings={"help_option_names": ["-h", "--help"]},
)
@click.version_option(package_name=PACKAGE_NAME)
@output_options(root=True)
def algokit() -> None:
    pass


algokit.add_command(init_command)
algokit.add_command(sandbox_group)
