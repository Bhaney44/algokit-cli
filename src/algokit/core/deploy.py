# 1. User can call algokit deploy to different networks
# 2. By default algokit cli contains configs for testnet and mainnet
# 3. User can overwrite them by creating a config file in the project root


import logging
import os
import sys
from pathlib import Path

import click
import httpx
from dotenv import dotenv_values

from algokit.core import proc
from algokit.core.conf import get_algokit_config
from algokit.core.constants import ALGOKIT_CONFIG, ALGORAND_NETWORKS, AlgorandNetworkConfiguration

if sys.version_info >= (3, 11):
    pass
else:
    pass

logger = logging.getLogger(__name__)


def get_genesis_network_name(deploy_config: dict) -> str | None:
    """
    Get the network name from the genesis block.
    :param deploy_config: Deploy configuration.
    :return: Network name.
    """

    algod_server = deploy_config.get("ALGOD_SERVER")

    if not algod_server:
        raise click.ClickException("Missing ALGOD_SERVER in deploy configuration.")

    try:
        genesis_response = httpx.get(f"{algod_server}/genesis")
        genesis_response.raise_for_status()
        return genesis_response.json()["network"] if genesis_response.status_code == 200 else None  # noqa: PLR2004
    except httpx.HTTPError:
        logger.warning(f"Failed to load network name from {algod_server} due to HTTP error.", exc_info=True)
        return None


def load_deploy_config(network: str, project_dir: Path) -> dict:
    """
    Load the deploy configuration for the given network.
    :param network: Network name.
    :param project_dir: Project directory path.
    :return: Deploy configuration.
    """

    env_file_path = project_dir / f".env.{network}"
    if not env_file_path.exists():
        if network in ALGORAND_NETWORKS:
            logger.info(f"No .env file found at {env_file_path}, loading default configuration for {network}.")
            return dict(ALGORAND_NETWORKS.get(network, {}))

        raise click.ClickException(f"No .env file found at {env_file_path}")

    # Extract network configuration
    network_config = dotenv_values(env_file_path, verbose=True, interpolate=True)

    # Fill missing keys with empty strings
    for required_key in list(AlgorandNetworkConfiguration.__annotations__.keys()):
        if required_key not in network_config:
            logger.warning(f"Warning: {required_key} is missing, filling with an empty string.")
            network_config[required_key] = ""

    return network_config


def load_deploy_command(network_name: str, project_dir: Path) -> str:
    """
    Load the deploy command for the given network from .algokit.toml file.
    :param network_name: Network name.
    :param project_dir: Project directory path.
    :return: Deploy command.
    """

    # Load and parse the TOML configuration file
    config = get_algokit_config(project_dir)

    if not config:
        raise click.ClickException(
            f"Couldn't load {ALGOKIT_CONFIG} file. Ensure deploy command is specified, either via "
            f"--custom-deploy-command or inside {ALGOKIT_CONFIG} file."
        )

    # Extract the deploy command for the given network
    try:
        return str(config["deploy"][network_name]["command"])
    except KeyError:
        raise click.ClickException(f"Deploy command is not specified in '{ALGOKIT_CONFIG}' file.") from None


def execute_deploy_command(command: str, deploy_config: dict, project_dir: Path | None) -> None:
    """
    Execute the deploy command.
    :param command: Deploy command.
    :param deploy_config: Deploy configuration.
    :param project_dir: Project directory path.
    """

    # Parse the deploy command
    try:
        # Get a copy of the current environment variables
        env = os.environ.copy()

        # Update the environment with custom variables from network_config
        env.update({key.upper(): str(value) for key, value in deploy_config.items()})

        proc.run(command.split(), env=env, cwd=project_dir or Path.cwd(), stdout_log_level=logging.INFO)
    except Exception as ex:
        raise click.ClickException(f"Failed to execute deploy command '{command}'.") from ex
