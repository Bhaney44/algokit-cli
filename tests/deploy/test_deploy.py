import os

import pytest
from _pytest.tmpdir import TempPathFactory
from algokit.cli.deploy import DEPLOYER_KEY, DISPENSER_KEY, LOCALNET, extract_env_variables
from algokit.core.constants import ALGOKIT_CONFIG
from approvaltests.namer import NamerFactory

from tests.utils.approvals import verify
from tests.utils.click_invoker import invoke


# Unit tests for extract_env_variables function
def test_extract_env_variables() -> None:
    # Set environment variables for the test
    os.environ[DEPLOYER_KEY] = "test_deployer_mnemonic"
    os.environ[DISPENSER_KEY] = "test_dispenser_mnemonic"

    # Test extraction with skip_mnemonics_prompts=True
    deployer_mnemonic, dispenser_mnemonic = extract_env_variables(True, LOCALNET)  # noqa: FBT003
    assert deployer_mnemonic == "test_deployer_mnemonic"
    assert dispenser_mnemonic == "test_dispenser_mnemonic"

    # Test extraction with skip_mnemonics_prompts=False and network != LOCALNET
    deployer_mnemonic, dispenser_mnemonic = extract_env_variables(False, "testnet")  # noqa: FBT003
    assert deployer_mnemonic == "test_deployer_mnemonic"
    assert dispenser_mnemonic == "test_dispenser_mnemonic"

    # Clean up environment variables
    del os.environ[DEPLOYER_KEY]
    del os.environ[DISPENSER_KEY]


# Approvals tests for deploy command
@pytest.mark.parametrize("network", ["", "localnet", "testnet", "mainnet", "betanet", "customnet"])
def test_deploy_command_no_algokit_toml(network: str, tmp_path_factory: TempPathFactory) -> None:
    cwd = tmp_path_factory.mktemp("cwd")

    result = invoke(
        f"deploy {network}",
        cwd=cwd,
    )

    assert result.exit_code == 1
    verify(
        result.output,
    )


@pytest.mark.parametrize("network", ["betanet", "localnet", "testnet", "mainnet"])
def test_deploy_command_default_networks_no_env(network: str, tmp_path_factory: TempPathFactory) -> None:
    cwd = tmp_path_factory.mktemp("cwd")
    if network != "localnet":
        os.environ[DEPLOYER_KEY] = "test_deployer_mnemonic"

    (cwd / ALGOKIT_CONFIG).write_text(f'[deploy.{network}]\ncommand = "echo HelloWorld"\n')

    input_answers = ["N" if network != "localnet" else ""]

    if network == "mainnet":
        input_answers.append("Y")

    result = invoke(
        f"deploy {network}",
        cwd=cwd,
        input="\n".join(input_answers),
    )

    assert result.exit_code == 0 if network == "localnet" else 1
    verify(result.output, options=NamerFactory.with_parameters(network))


@pytest.mark.parametrize("has_env", [True, False])
def test_deploy_command_custom_network_no_env(has_env: bool, tmp_path_factory: TempPathFactory) -> None:  # noqa: FBT001
    cwd = tmp_path_factory.mktemp("cwd")
    os.environ[DEPLOYER_KEY] = "test_deployer_mnemonic"
    network = "customnet"

    (cwd / ALGOKIT_CONFIG).write_text(f'[deploy.{network}]\ncommand = "echo HelloWorld"\n')

    if has_env:
        (cwd / f".env.{network}").write_text(
            """
            ALGOD_TOKEN=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
            ALGOD_SERVER=http://localhost
            ALGOD_PORT=4001
            INDEXER_TOKEN=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
            INDEXER_SERVER=http://localhost
            INDEXER_PORT=8980
            """
        )

    input_answers = ["N"]

    result = invoke(
        f"deploy {network}",
        cwd=cwd,
        input="\n".join(input_answers),
    )

    assert result.exit_code == 0 if network == "localnet" else 1
    verify(result.output, options=NamerFactory.with_parameters(has_env))
