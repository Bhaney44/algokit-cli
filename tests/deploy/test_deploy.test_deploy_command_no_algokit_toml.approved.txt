DEBUG: No .algokit.toml file found in the project directory.
Error: Couldn't load .algokit.toml file. Ensure deploy command is specified, either via --custom-deploy-command or inside .algokit.toml file.
